### ISSUE TITLE
Question about backprop 

### ISSUE THREAD

**Issue by edwhu** [nicklashansen/tdmpc#20]
Hi Nicklas, 

I am looking into this and the FOWM codebase, and have a few questions.

Q1: Backprop through time?

First, in the TDMPC codebase, it seems like you detach all the zs in the latent rollout. Doesn't this stop the backprop through time? If I do Q(z_3), the loss for Q(z_3) will not affect z_2, z_1.

https://github.com/nicklashansen/tdmpc/blob/f4d85eca7419039b71bab2234ffc8aca378dd313/src/algorithm/tdmpc.py#L194-L200

In the FOWM codebase, it  stores all the predicted zs into an array without detaching, and then predicts Q values for all the predicted zs. 
https://github.com/fyhMer/fowm/blob/bf7985876eb4baa827f41567cb6fb47b5da93ed4/src/algorithm/tdmpc.py#L283-L299

Q2: Why enable / disable tracking of Q gradients during update_pi?

If we are using the `pi_optim` to optimize, that would only update the policy parameters. So why do we need to disable / enable the Q network grads? Does that somehow make things more efficient? 

https://github.com/nicklashansen/tdmpc/blob/f4d85eca7419039b71bab2234ffc8aca378dd313/src/algorithm/tdmpc.py#L153-L169

**Comment by nicklashansen** [comment]
Hi @edwhu, sorry that I missed this!

1. The detach in L200 only affects the policy objective. We currently do not backprop through time when optimizing the policy. However, recent work (https://arxiv.org/abs/2407.02466) seems to indicate that there's room for improvement when it comes to the policy objective.
2. We disable Q-fn gradients for the sake of computational efficiency, yes. Another solution would be to compute gradients for Q and then reset them before next update of the Q-fn. Our implementation is directly inspired by SAC codebases such as https://github.com/denisyarats/pytorch_sac

I hope this helps!

**Comment by nicklashansen** [comment]
Also I should add that the separation of model and policy updates is made more explicit in the TD-MPC2 codebase, e.g. detach is called only *after* the model has been updated: https://github.com/nicklashansen/tdmpc2/blob/fabf01a5ec92e7cc5793ee9cbbced881a9c13798/tdmpc2/tdmpc2.py#L283
Hopefully that addresses any confusion!

**Comment by edwhu** [comment]
Thank you for the info!
